File: swift/tests/BedrockTests/BedrockToolingTests.swift
import XCTest

@testable import Bedrock

// Foreign Tests for tooling functionality (i.e. logging and error handling)
// The demo structs are only available in Foreign Tests and are not available in built binaries.

final class BedrockToolingTests: XCTestCase {

    func testToolingDemoLogPrefixing() throws {
        // Test the ToolingDemo to verify log prefixing works
        let demo = ToolingDemo()
        
        // These calls should generate logs with [Bedrock][ToolingDemo] prefix
        demo.logMessage(message: "Testing log prefixing from Swift")
        demo.testLogLevels()
        
        let result = demo.getDemoResult()
        XCTAssertTrue(result.contains("ToolingDemo"), "Result should contain the demo name")
        XCTAssertTrue(result.contains("Demo result"), "Result should contain expected text")
    }

    // MARK: - Error Handling Tests

    // Test: Strongly typed errors for validation and known cases
    func testDemoAuthenticate_StronglyTypedErrors() throws {
        let demo = ToolingDemo()

        // Success case - now includes result from post-auth operation
        let result = try demo.demoAuthenticate(username: "testuser", password: "validpassword")
        XCTAssertTrue(result.contains("Welcome, testuser!"))
        XCTAssertTrue(result.contains("Successfully processed: auth_data_testuser"))

        // Empty username - InvalidInput
        XCTAssertThrowsError(try demo.demoAuthenticate(username: "", password: "password")) { error in
            if let demoError = error as? DemoError,
                case .InvalidInput(let message) = demoError
            {
                XCTAssertTrue(message.contains("Username cannot be empty"))
            } else {
                XCTFail("Expected InvalidInput error")
            }
        }

        // Wrong credentials - AuthenticationFailed
        XCTAssertThrowsError(try demo.demoAuthenticate(username: "admin", password: "wrongpassword")) {
            error in
            if let demoError = error as? DemoError,
                case .AuthenticationFailed(let message) = demoError
            {
                XCTAssertTrue(message.contains("Authentication failed") && message.contains("401"))
            } else {
                XCTFail("Expected AuthenticationFailed error")
            }
        }

        // Slow user - NetworkTimeout
        XCTAssertThrowsError(try demo.demoAuthenticate(username: "slowuser", password: "password")) {
            error in
            if let demoError = error as? DemoError,
                case .NetworkTimeout(let message) = demoError
            {
                XCTAssertTrue(message.contains("Network timeout") && message.contains("30"))
            } else {
                XCTFail("Expected NetworkTimeout error")
            }
        }
    }

    // Test: Generic errors for complex anyhow error chains
    func testDemoGenericOperation_AnyhowChains() throws {
        let demo = ToolingDemo()

        // Success case
        let result = try demo.demoGenericOperation(input: "valid_input")
        XCTAssertEqual(result, "Successfully processed: valid_input")

        // Empty input - Generic error
        XCTAssertThrowsError(try demo.demoGenericOperation(input: "")) { error in
            if let demoError = error as? DemoError,
                case .Generic(let message) = demoError
            {
                XCTAssertTrue(message.contains("Input cannot be empty"))
            } else {
                XCTFail("Expected Generic error")
            }
        }

        // Network error - Generic error with anyhow context
        XCTAssertThrowsError(try demo.demoGenericOperation(input: "network_error")) { error in
            if let demoError = error as? DemoError,
                case .Generic(let message) = demoError
            {
                XCTAssertTrue(message.contains("Connection timed out"))
            } else {
                XCTFail("Expected Generic error")
            }
        }

        // Parse error - Generic error with anyhow context
        XCTAssertThrowsError(try demo.demoGenericOperation(input: "parse_error")) { error in
            if let demoError = error as? DemoError,
                case .Generic(let message) = demoError
            {
                XCTAssertTrue(message.contains("Failed to parse input as JSON"))
            } else {
                XCTFail("Expected Generic error")
            }
        }
    }

    // Test: Mixed usage - structured validation + generic processing
    func testDemoMixedOperation_CombinedApproach() throws {
        let demo = ToolingDemo()

        // Success case
        let result = try demo.demoMixedOperation(operation: "process", data: "valid_data")
        XCTAssertTrue(result.contains("Processed:"))

        // Empty operation - InvalidInput (strongly typed validation)
        XCTAssertThrowsError(try demo.demoMixedOperation(operation: "", data: "data")) { error in
            if let demoError = error as? DemoError,
                case .InvalidInput(let message) = demoError
            {
                XCTAssertTrue(message.contains("Operation cannot be empty"))
            } else {
                XCTFail("Expected InvalidInput error")
            }
        }

        // Unknown operation - InvalidInput (strongly typed validation)
        XCTAssertThrowsError(try demo.demoMixedOperation(operation: "unknown", data: "data")) { error in
            if let demoError = error as? DemoError,
                case .InvalidInput(let message) = demoError
            {
                XCTAssertTrue(message.contains("Unknown operation"))
            } else {
                XCTFail("Expected InvalidInput error")
            }
        }

        // Process operation with trigger_error - Generic error (anyhow processing)
        XCTAssertThrowsError(try demo.demoMixedOperation(operation: "process", data: "trigger_error")) {
            error in
            if let demoError = error as? DemoError,
                case .Generic(let message) = demoError
            {
                XCTAssertTrue(
                    message.contains("Operation failed")
                        && message.contains("Simulated processing failure"))
            } else {
                XCTFail("Expected Generic error with anyhow-style message")
            }
        }
    }

    // Test: Unified error handling across all demo functions
    func testUnifiedErrorHandling() {
        let demo = ToolingDemo()
        var caughtErrors: [DemoError] = []

        // Collect errors from different functions - all use same DemoError type
        do { _ = try demo.demoAuthenticate(username: "admin", password: "wrongpassword") } catch let
            error as DemoError
        { caughtErrors.append(error) } catch { XCTFail("Expected DemoError") }

        do { _ = try demo.demoGenericOperation(input: "network_error") } catch let error as DemoError {
            caughtErrors.append(error)
        } catch { XCTFail("Expected DemoError") }

        do { _ = try demo.demoMixedOperation(operation: "process", data: "trigger_error") } catch let
            error as DemoError
        { caughtErrors.append(error) } catch { XCTFail("Expected DemoError") }

        // Verify we have the three core error patterns
        XCTAssertEqual(caughtErrors.count, 3)

        // Strongly typed error
        if case .AuthenticationFailed = caughtErrors[0] {
        } else {
            XCTFail("Expected AuthenticationFailed")
        }

        // Generic error from anyhow chain
        if case .Generic = caughtErrors[1] {} else { XCTFail("Expected Generic") }

        // Generic error with prefix
        if case .Generic = caughtErrors[2] {} else { XCTFail("Expected Generic") }

        // All errors provide consistent localized descriptions
        for error in caughtErrors {
            XCTAssertFalse(error.localizedDescription.isEmpty)
        }
    }
    
    // MARK: - BedrockConfig Tests
    
    func testBedrockConfigInitialization() throws {
        // Initialize config with staging environment
        setConfig(environment: .staging)
        
        // Verify current environment is staging
        let config = getConfig()
        XCTAssertNotNil(config, "Config should be available after initialization")
        XCTAssertEqual(config?.environment(), .staging, "Environment should be staging after initialization")
        
        // Verify config is initialized
        XCTAssertTrue(isInitialized(), "Config should be initialized")
        
        // Get config and verify environment
        if let config = getConfig() {
            XCTAssertEqual(config.environment(), .staging, "Config environment should be staging")
        } else {
            XCTFail("Config should be available after initialization")
        }
        
        // Try to initialize again - should be ignored (check logs for warning)
        setConfig(environment: .production)
        
        // Environment should still be staging
        let configAfterSecondInit = getConfig()
        XCTAssertEqual(configAfterSecondInit?.environment(), .staging, "Environment should remain staging after second init attempt")
    }
    
    func testBedrockConfigEnvironmentTypes() throws {
        // Test creating config with different environments
        let stagingConfig = BedrockConfig(environment: .staging)
        XCTAssertEqual(stagingConfig.environment(), .staging, "Staging config should have staging environment")
        
        let productionConfig = BedrockConfig(environment: .production)
        XCTAssertEqual(productionConfig.environment(), .production, "Production config should have production environment")
    }
    
    // MARK: - Async Operation Tests
    
    func testDemoAsyncOperation_Success() async throws {
        let demo = ToolingDemo()
        
        // Test successful async operation with short delay
        let result = try await demo.demoAsyncOperation(delayMs: 100)
        XCTAssertTrue(result.contains("Async operation completed after 100ms"))
        XCTAssertTrue(result.contains("completed"))
    }
    
    func testDemoAsyncOperation_Timeout() async throws {
        let demo = ToolingDemo()
        
        // Test async operation that should timeout (over 5000ms)
        do {
            _ = try await demo.demoAsyncOperation(delayMs: 6000)
            XCTFail("Expected timeout error")
        } catch let error as DemoError {
            if case .Generic(let message) = error {
                XCTAssertTrue(message.contains("timeout exceeded"))
                XCTAssertTrue(message.contains("5 seconds"))
            } else {
                XCTFail("Expected Generic error for timeout")
            }
        }
    }
    
    func testDemoAsyncOperation_MultipleOperations() async throws {
        let demo = ToolingDemo()
        
        // Test multiple async operations to ensure runtime stability
        let result1 = try await demo.demoAsyncOperation(delayMs: 50)
        let result2 = try await demo.demoAsyncOperation(delayMs: 100)
        let result3 = try await demo.demoAsyncOperation(delayMs: 150)
        
        XCTAssertTrue(result1.contains("completed after 50ms"))
        XCTAssertTrue(result2.contains("completed after 100ms"))
        XCTAssertTrue(result3.contains("completed after 150ms"))
    }
    
    func testDemoAsyncOperation_ConcurrentOperations() async throws {
        // This test specifically verifies that the automatic tokio runtime configuration
        // added by bedrock_export works correctly with concurrent async operations in Swift
        let demo = ToolingDemo()
        
        // Run concurrent async operations to stress test the runtime
        let delays: [UInt64] = [10, 25, 50, 75, 100]
        
        let results = try await withThrowingTaskGroup(of: String.self) { group in
            for delay in delays {
                group.addTask {
                    return try await demo.demoAsyncOperation(delayMs: delay)
                }
            }
            
            var collectedResults: [String] = []
            for try await result in group {
                collectedResults.append(result)
            }
            return collectedResults
        }
        
        // Verify all operations completed successfully
        XCTAssertEqual(results.count, 5)
        for result in results {
            XCTAssertTrue(result.contains("completed"))
            XCTAssertTrue(result.contains("ms"))
        }
    }
    
    func testDemoAsyncOperation_RuntimeIntegration() async throws {
        // Additional test to verify sequential async operations work correctly
        let demo = ToolingDemo()
        
        let delays: [UInt64] = [20, 40, 60, 80, 100]
        var results: [String] = []
        
        for delay in delays {
            let result = try await demo.demoAsyncOperation(delayMs: delay)
            results.append(result)
        }
        
        // Verify all operations completed successfully
        XCTAssertEqual(results.count, 5)
        for (index, result) in results.enumerated() {
            let expectedDelay = delays[index]
            XCTAssertTrue(result.contains("completed after \(expectedDelay)ms"))
        }
    }
} 

File: kotlin/bedrock-tests/src/test/kotlin/bedrock/BedrockToolingTests.kt
package bedrock

import uniffi.bedrock.DemoException
import uniffi.bedrock.ToolingDemo
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import kotlinx.coroutines.runBlocking

// Foreign Tests for tooling functionality (i.e. logging and error handling)
// The demo structs are only available in Foreign Tests and are not available in built binaries.

class BedrockToolingTests {
    @Test
    fun testToolingDemoLogPrefixing() {
        // Test the ToolingDemo to verify log prefixing works
        val demo = ToolingDemo()

        // These calls should generate logs with [Bedrock][ToolingDemo] prefix
        demo.logMessage("Testing log prefixing from Kotlin")
        demo.testLogLevels()

        val result = demo.getDemoResult()
        assertTrue(result.contains("ToolingDemo"), "Result should contain the demo name")
        assertTrue(result.contains("Demo result"), "Result should contain expected text")
    }

    // MARK: - Error Handling Tests

    // Test: Strongly typed errors for validation and known cases
    @Test
    fun testDemoAuthenticate_StronglyTypedErrors() {
        val demo = ToolingDemo()

        // Success case - now includes result from post-auth operation
        val result = demo.demoAuthenticate("testuser", "validpassword")
        assertTrue(result.contains("Welcome, testuser!"))
        assertTrue(result.contains("Successfully processed: auth_data_testuser"))

        // Empty username - InvalidInput
        val emptyUsernameException =
            assertFailsWith<DemoException.InvalidInput> {
                demo.demoAuthenticate("", "password")
            }
        assertTrue(emptyUsernameException.message?.contains("Username cannot be empty") == true)

        // Wrong credentials - AuthenticationFailed
        val wrongCredentialsException =
            assertFailsWith<DemoException.AuthenticationFailed> {
                demo.demoAuthenticate("admin", "wrongpassword")
            }
        assertEquals(401u, wrongCredentialsException.code)
        assertTrue(wrongCredentialsException.message?.contains("401") == true)

        // Slow user - NetworkTimeout
        val timeoutException =
            assertFailsWith<DemoException.NetworkTimeout> {
                demo.demoAuthenticate("slowuser", "password")
            }
        assertEquals(30u, timeoutException.seconds)
        assertTrue(timeoutException.message?.contains("30") == true)
    }

    // Test: Generic errors for complex anyhow error chains
    @Test
    fun testDemoGenericOperation_AnyhowChains() {
        val demo = ToolingDemo()

        // Success case
        val result = demo.demoGenericOperation("valid_input")
        assertEquals("Successfully processed: valid_input", result)

        // Empty input - Generic error
        val emptyInputException =
            assertFailsWith<DemoException.Generic> {
                demo.demoGenericOperation("")
            }
        assertTrue(emptyInputException.message?.contains("Input cannot be empty") == true)

        // Network error - Generic error with anyhow context
        val networkException =
            assertFailsWith<DemoException.Generic> {
                demo.demoGenericOperation("network_error")
            }
        assertTrue(networkException.message?.contains("Connection timed out") == true)

        // Parse error - Generic error with anyhow context
        val parseException =
            assertFailsWith<DemoException.Generic> {
                demo.demoGenericOperation("parse_error")
            }
        assertTrue(parseException.message?.contains("Failed to parse input as JSON") == true)
    }

    // Test: Mixed usage - structured validation + generic processing
    @Test
    fun testDemoMixedOperation_CombinedApproach() {
        val demo = ToolingDemo()

        // Success case
        val result = demo.demoMixedOperation("process", "valid_data")
        assertTrue(result.contains("Processed:"))

        // Empty operation - InvalidInput (strongly typed validation)
        val emptyOperationException =
            assertFailsWith<DemoException.InvalidInput> {
                demo.demoMixedOperation("", "data")
            }
        assertTrue(emptyOperationException.message?.contains("Operation cannot be empty") == true)

        // Unknown operation - InvalidInput (strongly typed validation)
        val unknownOperationException =
            assertFailsWith<DemoException.InvalidInput> {
                demo.demoMixedOperation("unknown", "data")
            }
        assertTrue(unknownOperationException.message?.contains("Unknown operation") == true)

        // Process operation with trigger_error - Generic error (anyhow processing)
        val processingException =
            assertFailsWith<DemoException.Generic> {
                demo.demoMixedOperation("process", "trigger_error")
            }
        assertTrue(processingException.message?.contains("Operation failed") == true)
        assertTrue(processingException.message?.contains("Simulated processing failure") == true)
    }

    // Test: Unified error handling across all demo functions
    @Test
    fun testUnifiedErrorHandling() {
        val demo = ToolingDemo()
        val caughtErrors = mutableListOf<DemoException>()

        // Collect errors from different functions - all use same DemoException type
        try {
            demo.demoAuthenticate("admin", "wrongpassword")
        } catch (error: DemoException.AuthenticationFailed) {
            caughtErrors.add(error)
        }

        try {
            demo.demoGenericOperation("network_error")
        } catch (error: DemoException.Generic) {
            caughtErrors.add(error)
        }

        try {
            demo.demoMixedOperation("process", "trigger_error")
        } catch (error: DemoException.Generic) {
            caughtErrors.add(error)
        }

        // Verify we have the three core error patterns
        assertEquals(3, caughtErrors.size)

        // Strongly typed error
        assertTrue(caughtErrors[0] is DemoException.AuthenticationFailed, "Expected AuthenticationFailed")

        // Generic error from anyhow chain
        assertTrue(caughtErrors[1] is DemoException.Generic, "Expected Generic")

        // Generic error with prefix
        assertTrue(caughtErrors[2] is DemoException.Generic, "Expected Generic")

        // All errors provide consistent message content
        for (error in caughtErrors) {
            assertTrue(error.message?.isNotEmpty() == true, "Error message should not be empty")
        }
    }

    // MARK: - BedrockConfig Tests

    @Test
    fun testBedrockConfigInitialization() {
        // Initialize config with staging environment
        uniffi.bedrock.setConfig(uniffi.bedrock.BedrockEnvironment.STAGING, uniffi.bedrock.Os.ANDROID)

        // Verify current environment is staging
        val config = uniffi.bedrock.getConfig()
        assertNotNull(config, "Config should be available after initialization")
        assertEquals(uniffi.bedrock.BedrockEnvironment.STAGING, config.environment(), "Environment should be staging after initialization")

        // Verify config is initialized
        assertTrue(uniffi.bedrock.isInitialized(), "Config should be initialized")

        // Try to initialize again - should be ignored (check logs for warning)
        uniffi.bedrock.setConfig(uniffi.bedrock.BedrockEnvironment.PRODUCTION, uniffi.bedrock.Os.ANDROID)

        // Environment should still be staging
        val configAfterSecondInit = uniffi.bedrock.getConfig()
        assertEquals(
            uniffi.bedrock.BedrockEnvironment.STAGING,
            configAfterSecondInit?.environment(),
            "Environment should remain staging after second init attempt",
        )
    }

    @Test
    fun testBedrockConfigEnvironmentTypes() {
        // Test creating config with different environments
        val stagingConfig = uniffi.bedrock.BedrockConfig(uniffi.bedrock.BedrockEnvironment.STAGING, uniffi.bedrock.Os.ANDROID)
        assertEquals(
            uniffi.bedrock.BedrockEnvironment.STAGING,
            stagingConfig.environment(),
            "Staging config should have staging environment",
        )

        val productionConfig = uniffi.bedrock.BedrockConfig(uniffi.bedrock.BedrockEnvironment.PRODUCTION, uniffi.bedrock.Os.ANDROID)
        assertEquals(
            uniffi.bedrock.BedrockEnvironment.PRODUCTION,
            productionConfig.environment(),
            "Production config should have production environment",
        )
    }

    // MARK: - Async Operation Tests
    
    @Test
    fun testDemoAsyncOperation_Success() = runBlocking {
        val demo = ToolingDemo()
        
        // Test successful async operation with short delay
        val result = demo.demoAsyncOperation(100uL)
        assertTrue(result.contains("Async operation completed after 100ms"))
        assertTrue(result.contains("completed"))
    }
    
    @Test
    fun testDemoAsyncOperation_Timeout() = runBlocking {
        val demo = ToolingDemo()
        
        // Test async operation that should timeout (over 5000ms)
        val timeoutException = assertFailsWith<DemoException.Generic> {
            demo.demoAsyncOperation(6000uL)
        }
        assertTrue(timeoutException.message?.contains("timeout exceeded") == true)
        assertTrue(timeoutException.message?.contains("5 seconds") == true)
    }
    
    @Test
    fun testDemoAsyncOperation_MultipleOperations() = runBlocking {
        val demo = ToolingDemo()
        
        // Test multiple async operations to ensure runtime stability
        val result1 = demo.demoAsyncOperation(50uL)
        val result2 = demo.demoAsyncOperation(100uL)
        val result3 = demo.demoAsyncOperation(150uL)
        
        assertTrue(result1.contains("completed after 50ms"))
        assertTrue(result2.contains("completed after 100ms"))
        assertTrue(result3.contains("completed after 150ms"))
    }
    
    @Test
    fun testDemoAsyncOperation_RuntimeIntegration() = runBlocking {
        // This test specifically verifies that the automatic tokio runtime configuration
        // added by bedrock_export works correctly in foreign code
        val demo = ToolingDemo()
        
        // Run a series of async operations to stress test the runtime
        val delays = listOf(10uL, 25uL, 50uL, 75uL, 100uL)
        val results = mutableListOf<String>()
        
        for (delay in delays) {
            val result = demo.demoAsyncOperation(delay)
            results.add(result)
        }
        
        // Verify all operations completed successfully
        assertEquals(5, results.size)
        for ((index, result) in results.withIndex()) {
            val expectedDelay = delays[index]
            assertTrue(result.contains("completed after ${expectedDelay}ms"))
        }
    }
}
